src/Analyzers/CSharp/CodeFixes/RemoveUnnecessarySuppressions/CSharpRemoveUnnecessaryNullableWarningSuppressionsCodeFixProvider.cs

```diff
@@ -3,50 +3,178 @@
+            var filterBasedOnScope = fixAllContext.Scope == FixAllScope.Solution;
+
+            // Map from a document to all linked documents it has (not including itself).
+            using var _ = PooledDictionary<DocumentId, ImmutableArray<DocumentId>>.GetInstance(out var documentToLinkedDocuments);
+
+            PopulateLinkedDocumentMap();
+            var updatedSolution = await ProcessLinkedDocumentMapAsync().ConfigureAwait(false);
+
+            return CodeAction.Create(
+                fixAllContext.GetDefaultFixAllTitle(),
+                (_, _) => Task.FromResult(updatedSolution),
+                equivalenceKey: null,
+                CodeActionPriority.Default
+#if !CODE_STYLE
+                , this.Cleanup
+#endif
+                );
+
+            void PopulateLinkedDocumentMap()
+            {
+                var solution = fixAllContext.Solution;
+                foreach (var (document, _) in documentToDiagnostics)
+                {
+                    // Note: GetLinkedDocuments does not return the document it was called on.
+                    var linkedDocuments = document.GetLinkedDocumentIds();
+
+                    // Ignore any linked documents we already saw by processing another document in that linked set.
+                    if (linkedDocuments.Any(id => documentToLinkedDocuments.ContainsKey(id)))
+                        continue;
+
+                    documentToLinkedDocuments[document.Id] = linkedDocuments;
+                }
+            }
+
+            async Task<Solution> ProcessLinkedDocumentMapAsync()
+            {
+                var currentSolution = fixAllContext.Solution;
+
+                foreach (var (documentId, linkedDocumentIds) in documentToLinkedDocuments)
+                {
+                    // Now, for each group of linked documents, only remove the suppression operators we see in all documents.
+                    var document = fixAllContext.Solution.GetRequiredDocument(documentId);
+                    using var _ = PooledHashSet<TextSpan>.GetInstance(out var commonSpans);
+
+                    var diagnostics = documentToDiagnostics[document];
+
+                    // Start initially with all the spans in this document.
+                    commonSpans.UnionWith(GetDiagnosticSpans(diagnostics));
+
+                    // Now, only keep those spans that are also in all other linked documents.
+                    if (filterBasedOnScope)
+                    {
+                        foreach (var linkedDocumentId in linkedDocumentIds)
+                        {
+                            var linkedDocument = fixAllContext.Solution.GetRequiredDocument(linkedDocumentId);
+                            var linkedDiagnostics = documentToDiagnostics.TryGetValue(linkedDocument, out var result) ? result : [];
+
+                            commonSpans.IntersectWith(GetDiagnosticSpans(linkedDiagnostics));
```

Comment:
is intersection correct - or should the spans be exactly the same? Thinking if there could be some weird case where directives cause two different TFMs to have intersecting diagnostics for different suppressions

Reply:
I personally thinkk intersection is correct.  these are the suppressions that both sides think are safe to remove.  I believe (though don't know if i could truly prove) that suppressions operate with subset semantics.  so if N of them are deemed unnecessary, you could remove any subset of those and be ok.